### ISSUE TITLE
Support for per-document attention masking

### ISSUE THREAD

**Issue by paul-gibbons** [NVIDIA-NeMo/Megatron-Bridge#902]
Add support for per-document attention masking as described below.

[Llama3](https://ai.meta.com/blog/meta-llama-3/) uses a mask to ensure self-attention does not cross document boundaries. 
[FlexAttention](https://pytorch.org/blog/flexattention/) has intuitive examples for how to use document masking.
